### PR TITLE
FAI-494 - adding impact-score LIME optimizer

### PR DESCRIPTION
**Pull request by tteofili**
See https://issues.redhat.com/browse/FAI-494
Many thanks for submitting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](https://github.com/kiegroup/kogito-runtimes#contributing-to-kogito)
- [ ] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [x] Pull Request title contains the target branch if not targeting master: `[0.9.x] KOGITO-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>Pull Request</b>  
  Please add comment: <b>Jenkins retest this</b>
 
* <b>Quarkus LTS checks</b>  
  Please add comment: <b>Jenkins run LTS</b>

* <b>Native checks</b>  
  Please add comment: <b>Jenkins run native</b>

</details>

### CODE REVIEW

**Comment by tteofili** [comment]
Jenkins retest this please

**Review by ruivieira — APPROVED** [review]
@tteofili LGTM, thanks!

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/lime/optim/LimeImpactScoreCalculator.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                    List<FeatureImportance> topFeatures = entry.getValue().getTopFeatures(TOP_FEATURES);
+                    if (!topFeatures.isEmpty()) {
+                        double v = ExplainabilityMetrics.impactScore(solution.getModel(),
+                                prediction, topFeatures);
+                        impactScore = impactScore.add(BigDecimal.valueOf(v));
+                        succeededEvaluations++;
+                    }
+                }
+            } catch (ExecutionException e) {
+                LOGGER.error("Saliency impact-score calculation returned an error {}", e.getMessage());
```
Comment (jiripetrlik) [inline explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/lime/optim/LimeImpactScoreCalculator.java:72]: Is it intentional to only log the error. Shouldn't we rather rethrow the exception?

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/lime/optim/LimeImpactScoreCalculator.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+import org.kie.kogito.explainability.utils.ExplainabilityMetrics;
+import org.optaplanner.core.api.score.buildin.simplebigdecimal.SimpleBigDecimalScore;
+import org.optaplanner.core.api.score.calculator.EasyScoreCalculator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LimeImpactScoreCalculator implements EasyScoreCalculator<LimeConfigSolution, SimpleBigDecimalScore> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(LimeImpactScoreCalculator.class);
+    private static final int TOP_FEATURES = 2;
```
Comment (jiripetrlik) [inline explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/lime/optim/LimeImpactScoreCalculator.java:40]: Should the number of top features be always 2 or should we rather parameterize?

File: explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/lime/optim/LimeImpactScoreCalculatorTest.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        LimeImpactScoreCalculator scoreCalculator = new LimeImpactScoreCalculator();
+        LimeConfig config = new LimeConfig();
+        List<Prediction> predictions = Collections.emptyList();
+        List<LimeConfigEntity> entities = Collections.emptyList();
+        PredictionProvider model = TestUtils.getDummyTextClassifier();
+        LimeConfigSolution solution = new LimeConfigSolution(config, predictions, entities, model);
+        SimpleBigDecimalScore score = scoreCalculator.calculateScore(solution);
+        assertThat(score).isNotNull();
+        assertThat(score.getScore()).isNotNull();
+        assertThat(score.getScore()).isEqualTo(BigDecimal.valueOf(0));
```
Comment (jiripetrlik) [inline explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/lime/optim/LimeImpactScoreCalculatorTest.java]: Can we add a test also for some example where score is not equal to zero?

**Review by jiripetrlik — COMMENTED** [review]
Hello @tteofili, thank you for your PR. I have few comments/questions.

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/lime/optim/LimeImpactScoreCalculator.java (hunk 1 above)
Comment (tteofili) [reply explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/lime/optim/LimeImpactScoreCalculator.java:72]: the intention here is not to block the optimization because of one single error only in generating the predictions (transient errors, timeouts in HTTP calls, etc.). 
Of course if all the executions fail then the optimization will not generate a different config.

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/lime/optim/LimeImpactScoreCalculator.java (hunk 2 above)
Comment (tteofili) [reply explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/lime/optim/LimeImpactScoreCalculator.java:40]: two is valid default which should work well for most of the models/decision services. 
At the moment I tend to think that this might become a parameter in the future if we see there're people interested in optimizing at a different granularity.

File: explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/lime/optim/LimeImpactScoreCalculatorTest.java (hunk 3 above)
Comment (tteofili) [reply explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/lime/optim/LimeImpactScoreCalculatorTest.java]: ok

**Comment by tteofili** [comment]
@jiripetrlik I think I've addressed your comments, please have a look when you have time.

**Review by jiripetrlik — APPROVED** [review]

**Review by danielezonca — APPROVED** [review]
